src/read/macinfo.rs

```diff
@@ -0,0 +1,290 @@
+            }
+        } else {
+            // If the offset is invalid, the iterator is immediately finished.
+            DebugMacInfoIterator {
+                input_reader: input,
+                finished: true,
+            }
+        }
+    }
+}
+
+impl<T> DebugMacInfo<T> {
+    /// Create a `DebugMacInfo` section that references the data in `self`.
+    ///
+    /// This is useful when `R` implements `Reader` but `T` does not.
+    ///
+    /// Used by `DwarfSections::borrow`.
+    pub fn borrow<'a, F, R>(&'a self, mut borrow: F) -> DebugMacInfo<R>
+    where
+        F: FnMut(&'a T) -> R,
+    {
+        borrow(&self.section).into()
+    }
+}
+
+impl<R> Section<R> for DebugMacInfo<R> {
+    fn id() -> SectionId {
+        SectionId::DebugMacinfo
+    }
+
+    fn reader(&self) -> &R {
+        &self.section
+    }
+}
+
+impl<R> From<R> for DebugMacInfo<R> {
+    fn from(macinfo_section: R) -> Self {
+        DebugMacInfo {
+            section: macinfo_section,
+        }
+    }
+}
+
+/// an Entry in the `.debug_macinfo` section.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum DebugMacInfoItem<R> {
+    /// The end of the unit.
+    EndOfUnit,
+    /// A macro definition.
+    Define {
+        /// The line number where the macro is defined.
+        line_number: u64,
+        /// The text of the macro: a null-terminated string.
+        macro_text: R,
+    },
+    /// A macro undefinition.
+    Undef {
+        /// The line number where the macro is undefined.
+        line_number: u64,
+        /// The text of the macro: a null-terminated string.
+        macro_text: R,
+    },
+    /// The start of a file.
+    StartFile {
+        /// line number of the source line on which the inclusion macro directive occurred
+        line_number: u64,
+        /// the index into the line number table of the compilation unit (LEB128).
+        file_number_index: u64,
+    },
+    /// The end of a file.
+    EndFile,
+    /// A vendor-specific extension.
+    VendorExt {
+        /// a numeric constant (LEB128), whose meaning is vendor specific.
+        numeric_constant: u64,
+        /// a null-terminated string, whose meaning is vendor specific.
+        string_constant: R,
+    },
+}
+
+/// Iterator over the entries in the `.debug_macinfo` section.
+///
+/// the iterator is not public and does not appear in the API - callers only get an `impl Iterator<Item = DebugMacInfoItem<R>>`
+struct DebugMacInfoIterator<R: Reader> {
+    input_reader: R,
+    finished: bool,
+}
+
+impl<R: Reader> Iterator for DebugMacInfoIterator<R> {
+    type Item = DebugMacInfoItem<R>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // make sure the iterator never returns Some(x) after returning None
+        if self.finished {
+            return None;
+        }
+        // Read the next entry from the input reader and return it as a DebugMacInfoItem.
+        let macinfo_type = DwMacInfo(self.input_reader.read_u8().ok()?);
```

Comment:
Please make this a fallible iterator instead. Never do `.ok()?`.

Reply:
All uses of fallible iterator depend on the "fallible-iterator" feature.
If I make this iterator a FallibleIterator, then reading the content of .debug_macinfo as a whole would have to depend on this feature. I avoided this dependency by invalidating the iterator on error, but if you think making it fallible is better then I'll change it.